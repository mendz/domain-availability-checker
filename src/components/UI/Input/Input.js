import React, { useState } from 'react';
import PropTypes from 'prop-types';

import classes from './Input.module.css';

import { ReactComponent as OpenEye } from '../../../assets/icons/_ionicons_svg_md-eye.svg';
import { ReactComponent as CloseEye } from '../../../assets/icons/_ionicons_svg_md-eye-off.svg';

const Input = ({
  label,
  elementConfig: { type, placeholder },
  changed,
  invalid,
  touched,
  value,
  errorMessage,
}) => {
  const [isFocus, setIsFocus] = useState(false); // to show the error only when the input us un-focus
  const [showPassword, setShowPassword] = useState(false);
  const classArr = [classes.Input];

  if (invalid && touched && !isFocus) {
    classArr.push(classes.Invalid);
  }

  let validationError = null;
  if (invalid && touched && !isFocus) {
    const errorMsg = errorMessage
      ? errorMessage
      : `Please enter valid ${label}`;
    validationError = <p className={classes.ValidationError}>{errorMsg}</p>;
  }

  let showPassButton = null;
  if (type === 'password') {
    showPassButton = (
      <button
        className={classes.ShowPasswordButton}
        onClick={() => setShowPassword(!showPassword)}
      >
        {!showPassword ? (
          <OpenEye height="1.2rem" width="1.2rem" />
        ) : (
          <CloseEye height="1.2rem" width="1.2rem" />
        )}
      </button>
    );
  }

  return (
    <div className={classes.Container}>
      <label htmlFor={type} className={classes.Label}>
        {label}
      </label>
      <div className={classes.InputWrapper}>
        <input
          type={showPassword ? 'text' : type}
          placeholder={placeholder}
          name={type}
          id={type}
          className={classArr.join(' ')}
          onChange={changed}
          value={value}
          onBlur={() => setIsFocus(false)}
          onFocus={() => setIsFocus(true)}
        />
        {showPassButton}
      </div>
      {validationError}
    </div>
  );
};

Input.propTypes = {
  label: PropTypes.string.isRequired,
  elementConfig: PropTypes.shape({
    type: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
  }).isRequired,
  changed: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  touched: PropTypes.bool.isRequired,
  value: PropTypes.string.isRequired,
  errorMessage: PropTypes.string,
};

export default Input;
