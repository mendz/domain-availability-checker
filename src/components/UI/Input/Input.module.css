.Container {
  width: 100%;
}

.Input {
  margin: 10px 0;
  width: 100%;
  background: #fff;
  color: var(--gray);
  border: 0;
  padding: 1.3rem 1rem;
  font-size: 1.1rem;
  border-radius: 5px;
  box-shadow: 1px 2px 3px 0 rgba(1, 1, 1, 0.25);
  transition: box-shadow 0.2s ease-in;
}

.InputWrapper {
  display: flex;
  align-items: center;
  position: relative;
}

.ShowPasswordButton {
  position: absolute;
  right: 2px;
  background: transparent;
  border: 0;
  cursor: pointer;
  border: 0;
}

.ShowPasswordButton:disabled {
  cursor: not-allowed;
}

.Input:focus {
  box-shadow: 1px 2px 3px 4px rgba(1, 1, 1, 0.25);
}

.Input:disabled {
  cursor: not-allowed;
  background: var(--gray);
}

.Label {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.Invalid {
  border: 1px solid var(--red);
  background-color: #fda49a;
}

.ValidationError {
  color: var(--red);
  margin: 5px 0;
  text-align: left;
  margin-bottom: 1rem;
}
