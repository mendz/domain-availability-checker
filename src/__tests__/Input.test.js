import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

import Input from '../components/UI/Input/Input';
import { ReactComponent as OpenEye } from '../assets/icons/_ionicons_svg_md-eye.svg';
import { ReactComponent as CloseEye } from '../assets/icons/_ionicons_svg_md-eye-off.svg';

describe('<Input />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Input
        label="label"
        elementConfig={{
          type: 'text',
          placeholder: 'placeholder',
        }}
        changed={() => {}}
        invalid={false}
        touched={false}
        value=""
        errorMessage={null}
      />
    );
  });

  it('matches the snapshot', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should display en error message if the invalid and touched values are true and the input is blur', () => {
    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('blur');
    wrapper.setProps({
      invalid: true,
      touched: true,
      errorMessage: 'some error message',
    });

    expect(wrapper.contains('some error message')).toBe(true);

    wrapper.setProps({
      invalid: false,
      touched: false,
      errorMessage: 'some error message',
    });

    expect(wrapper.contains('some error message')).toBe(false);

    wrapper.setProps({
      label: 'Email',
      invalid: true,
      touched: true,
      errorMessage: null,
    });

    expect(wrapper.contains('Please enter valid Email')).toBe(true);
  });

  it('should not display the error message until the input is blur', () => {
    wrapper.find('input').simulate('focus');
    wrapper.setProps({
      invalid: true,
      touched: true,
      errorMessage: 'some error message',
    });

    expect(wrapper.contains('some error message')).toBe(false);

    wrapper.setProps({
      label: 'Email',
      invalid: true,
      touched: true,
      errorMessage: null,
    });

    expect(wrapper.contains('Please enter valid Email')).toBe(false);

    wrapper.find('input').simulate('blur');

    expect(wrapper.contains('Please enter valid Email')).toBe(true);
  });

  it('should have show password button when the type is "password"', () => {
    expect(wrapper.exists('button.ShowPasswordButton')).toBe(false);
    wrapper.setProps({ elementConfig: { type: 'password' } });
    expect(wrapper.exists('button.ShowPasswordButton')).toBe(true);
  });

  it('should change the input password type to "text" and back to "password"', () => {
    wrapper.setProps({ elementConfig: { type: 'password' } });
    const showPasswordButton = wrapper.find('button.ShowPasswordButton');

    expect(wrapper.find('input').props().type).toBe('password');
    showPasswordButton.simulate('click');
    expect(wrapper.find('input').props().type).toBe('text');
  });

  it('The ShowPasswordButton button should have the correct svg eye depends the input type', () => {
    wrapper.setProps({ elementConfig: { type: 'password' } });
    const showPasswordButton = wrapper.find('button.ShowPasswordButton');

    expect(wrapper.containsMatchingElement(<OpenEye />)).toBe(true);
    showPasswordButton.simulate('click');
    expect(wrapper.containsMatchingElement(<CloseEye />)).toBe(true);
  });
});
