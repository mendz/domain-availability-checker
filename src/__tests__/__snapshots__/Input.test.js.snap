// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> matches the snapshot 1`] = `
<div
  className="Container"
>
  <label
    className="Label"
    htmlFor="text"
  >
    label
  </label>
  <div
    className="InputWrapper"
  >
    <input
      className="Input"
      id="text"
      name="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="placeholder"
      type="text"
      value=""
    />
  </div>
</div>
`;
